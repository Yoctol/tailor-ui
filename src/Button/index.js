import PropTypes from 'prop-types';
import React, { PureComponent, createRef } from 'react';
import styled, { css, keyframes } from 'styled-components';
import { ifProp, switchProp } from 'styled-tools';
import { space, themeGet } from 'styled-system';

import controlTransition from '../utils/transition';

import Ripple from './Ripple';

const spin = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`;

const StyledButton = styled.button`
  display: inline-flex;
  position: relative;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  border: ${themeGet('borders.default')};
  border-radius: ${themeGet('radii.1')};
  border-color: ${themeGet('colors.primary')};
  background-color: ${themeGet('colors.primaryDark')};
  color: ${themeGet('colors.light')};
  line-height: ${themeGet('lineHeight')};
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;

  &:focus {
    border-color: ${themeGet('colors.primaryDark')};
    outline: 0;
  }

  &:hover {
    background-color: ${themeGet('colors.light')};
    color: ${themeGet('colors.primaryDark')};
  }

  &:disabled,
  &[disabled] {
    opacity: 0.5;
    cursor: default;
    pointer-events: none;
  }

  ${switchProp('size', {
    sm: css`
      padding: ${themeGet('space.paddingYSm')} ${themeGet('space.paddingXSm')};
      font-size: ${themeGet('fontSizes.sm')};
    `,
    m: css`
      padding: ${themeGet('space.paddingY')} ${themeGet('space.paddingX')};
      font-size: ${themeGet('fontSizes.default')};
    `,
    lg: css`
      padding: ${themeGet('space.paddingYLg')} ${themeGet('space.paddingXLg')};
      font-size: ${themeGet('fontSizes.lg')};
    `,
  })}

  ${ifProp(
    'block',
    css`
      width: 100%;
    `
  )}

  ${ifProp(
    'danger',
    css`
      border-color: ${themeGet('colors.error')};
      background-color: ${themeGet('colors.error')};
      color: ${themeGet('colors.light')};

      &:focus {
        border-color: ${themeGet('colors.error')};
      }

      &:hover {
        background-color: ${themeGet('colors.error')};
        color: ${themeGet('colors.light')};
      }
    `
  )};

  ${ifProp(
    'light',
    css`
      border-color: ${themeGet('colors.gray.8')};
      background-color: ${themeGet('colors.light')};
      color: ${themeGet('colors.gray.2')};

      &:focus {
        border-color: ${themeGet('colors.primaryDark')};
      }

      &:hover {
        background-color: ${themeGet('colors.gray.8')};
      }

      ${ifProp(
        'active',
        css`
          border-color: ${themeGet('colors.secondaryDark')};
          background-color: ${themeGet('colors.secondaryDark')};
          color: ${themeGet('colors.light')};

          &:hover {
            background-color: ${themeGet('colors.secondary')};
            color: ${themeGet('colors.light')};
          }

          &:focus {
            border-color: ${themeGet('colors.secondaryDark')};
          }
        `
      )};
    `
  )};

  ${ifProp(
    'circle',
    css`
    border-radius: 999px;

    ${ifProp(
      { size: 'sm' },
      css`
        padding: ${themeGet('space.paddingYSm')}
          calc(${themeGet('space.paddingXSm')} * 2);
      `
    )}
    ${ifProp(
      { size: 'm' },
      css`
        padding: ${themeGet('space.paddingY')}
          calc(${themeGet('space.paddingX')} * 2);
      `
    )}
    ${ifProp(
      { size: 'lg' },
      css`
        padding: ${themeGet('space.paddingYLg')}
          calc(${themeGet('space.paddingXLg')} * 2);
      `
    )}
  `
  )}

  ${ifProp(
    'ghost',
    css`
      border-color: ${themeGet('colors.light')};
      background-color: transparent;
    `
  )}

  ${ifProp(
    'loading',
    css`
      color: transparent;
      pointer-events: none;

      &::after {
        content: '';
        display: block;
        position: absolute;
        top: calc(50% - (1em / 2));
        left: calc(50% - (1em / 2));
        width: 1em;
        height: 1em;
        border: 2px solid ${themeGet('colors.gray.6')};
        border-radius: 50%;
        border-top-color: transparent;
        border-right-color: transparent;
        animation: ${spin} 0.7s linear infinite;
      }
    `
  )}

  ${controlTransition()};

  ${space};
`;

class Button extends PureComponent {
  constructor(props) {
    super(props);
    this.ripple = createRef();
    this.button = createRef();
  }

  handleClick = e => this.ripple.current.startRipple(e, this.button);

  render() {
    const { children, ...props } = this.props;
    return (
      <StyledButton
        innerRef={this.button}
        onMouseUp={this.handleClick}
        onTouchend={this.handleClick}
        {...props}
      >
        {children}
        <Ripple ref={this.ripple} />
      </StyledButton>
    );
  }
}

Button.displayName = 'Button';

Button.propTypes = {
  /**
   * Description of prop "active".
   */
  active: PropTypes.bool,
  /**
   * Description of prop "block".
   */
  block: PropTypes.bool,
  /**
   * Description of prop "circle".
   */
  circle: PropTypes.bool,
  /**
   * Description of prop "ghost".
   */
  danger: PropTypes.bool,
  /**
   * Description of prop "light".
   */
  ghost: PropTypes.bool,
  /**
   * Description of prop "danger".
   */
  light: PropTypes.bool,
  /**
   * Description of prop "loading".
   */
  loading: PropTypes.bool,
  /**
   * Description of prop "size".
   */
  size: PropTypes.oneOf(['sm', 'm', 'lg']),
  /**
   * Description of prop "Space".
   */
  ...space.propTypes,
};

Button.defaultProps = {
  size: 'm',
  block: false,
  circle: false,
  danger: false,
  ghost: false,
  light: false,
  active: false,
  loading: false,
};

export default Button;
