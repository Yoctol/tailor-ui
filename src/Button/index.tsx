import React, {
  MouseEvent,
  PureComponent,
  ReactInstance,
  RefObject,
  createRef,
} from 'react';
import { SpaceProps, space as styledSpace } from 'styled-system';

import styled, { css, keyframes } from 'utils/styled-components';
import tag from 'utils/CleanTag';

import Icon, { IconType, IconWrapper } from '../Icon';

import Ripple from './Ripple';

const spin = keyframes`
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
`;

type ButtonType =
  | 'default'
  | 'primary'
  | 'secondary'
  | 'info'
  | 'success'
  | 'warning'
  | 'danger';

export interface IGetTypesStylesInterface {
  type?: ButtonType;
  text?: boolean;
  outlined?: boolean;
}

const getTypeStyles = ({
  type = 'default',
  text,
  outlined,
  theme,
}: IGetTypesStylesInterface & {
  theme: any;
}) => {
  const { colors } = theme;

  if (text) {
    return {
      borderColor: 'transparent',
      backgroundColor: 'transparent',
      color: colors.gray800,
      hover: {
        backgroundColor: colors.gray300,
        borderColor: 'transparent',
      },
    };
  }

  if (type === 'default') {
    return {
      borderColor: colors.gray300,
      backgroundColor: colors.light,
      color: colors.gray800,
      hover: {
        backgroundColor: colors.gray200,
        borderColor: colors.primary,
      },
    };
  }

  const bg = colors[type];
  const { light } = colors;
  const backgroundColor = outlined ? 'transparent' : bg;

  return {
    borderColor: bg,
    backgroundColor,
    color: outlined ? bg : light,
    hover: {
      backgroundColor: outlined ? bg : light,
      color: outlined ? light : bg,
    },
  };
};

interface IGetLoadingInterface extends IGetTypesStylesInterface {
  loading?: boolean;
}

const getLoading = ({
  loading,
  ...props
}: IGetLoadingInterface & {
  theme: any;
}) =>
  loading &&
  css`
    opacity: 0.7;
    cursor: default;
    pointer-events: none;

    &::before {
      width: 1em;
      height: 1em;
      margin-right: 8px;
      border: 1px solid ${getTypeStyles(props).color};
      border-radius: 50%;
      border-top-color: transparent;
      border-right-color: transparent;
      animation: ${spin} 0.7s linear infinite;
    }
  `;

type ButtonSize = 'sm' | 'md' | 'lg';

interface IGetSizeInterface {
  size: ButtonSize;
  hasIcon?: boolean;
}

const getSize = css`
  ${({
    hasIcon,
    size,
    theme: { paddings, heights, fontSizes },
  }: IGetSizeInterface & { theme: any }) => {
    if (hasIcon) {
      return {
        sm: css`
          width: ${heights.sm};
          height: ${heights.sm};
        `,
        md: css`
          width: ${heights.base};
          height: ${heights.base};
        `,
        lg: css`
          width: ${heights.lg};
          height: ${heights.lg};
        `,
      }[size];
    }

    return {
      sm: css`
        height: ${heights.sm};
        padding: 0 ${paddings.sm};
        font-size: ${fontSizes.sm};
      `,
      md: css`
        height: ${heights.base};
        padding: 0 ${paddings.md};
        font-size: ${fontSizes.base};
      `,
      lg: css`
        height: ${heights.lg};
        padding: 0 ${paddings.lg};
        font-size: ${fontSizes.lg};
      `,
    }[size];
  }};
`;

interface IBlockInterface {
  block?: boolean;
}

const getBlock = ({ block }: IBlockInterface) =>
  block &&
  css`
    width: 100%;
  `;

interface IRoundedInterface {
  rounded?: boolean;
}

const getRounded = ({ rounded }: IRoundedInterface) =>
  rounded &&
  css`
    border-radius: 999px;
  `;

const getTypes = ({
  type = 'default',
  text = false,
  outlined = false,
  theme,
}: IGetTypesStylesInterface & { theme: any }) => {
  const styles = getTypeStyles({ type, text, outlined, theme });

  return css`
    border-color: ${styles.borderColor};
    background-color: ${styles.backgroundColor};
    color: ${styles.color};

    ${IconWrapper /* sc-selector */} {
      cursor: pointer;
      svg {
        fill: ${styles.color};
      }
    }

    &:hover {
      border-color: ${styles.hover.borderColor};
      background-color: ${styles.hover.backgroundColor};
      color: ${styles.hover.color};

      ${IconWrapper /* sc-selector */} svg {
        fill: ${styles.hover.color};
      }
    }
  `;
};

type StyledButtonProps = IGetTypesStylesInterface &
  IGetLoadingInterface &
  IGetSizeInterface &
  IBlockInterface &
  IRoundedInterface &
  SpaceProps;

export const StyledButton = styled<StyledButtonProps, 'button'>(tag.button)`
  display: inline-flex;
  position: relative;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  line-height: ${p => p.theme.lineHeight};
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  user-select: none;

  &:focus {
    outline: 0;
  }

  &:disabled,
  &[disabled] {
    opacity: 0.7;
    cursor: default;
    pointer-events: none;
  }

   &::before {
      content: '';
      display: inline-block;
      position: relative;
      left: 0;
      width: 0;
      height: 0;
      margin-right: 0;
      ${p => p.theme.transition /* sc-declaration */};
    }

  ${getBlock /* sc-declaration */}
  ${getRounded /* sc-declaration */}
  ${getSize /* sc-declaration */}
  ${getTypes /* sc-declaration */}
  ${getLoading /* sc-declaration */};

  ${p => p.theme.transition /* sc-declaration */};
  ${styledSpace};
`;

export type ButtonProps = SpaceProps & {
  type?: ButtonType;
  text?: boolean;
  block?: boolean;
  rounded?: boolean;
  outlined?: boolean;
  loading?: boolean;
  icon?: string | IconType;
  size?: ButtonSize;
  onClick?: (event: MouseEvent) => void;
};

class Button extends PureComponent<ButtonProps> {
  ripple: RefObject<Ripple> = createRef<Ripple>();

  buttonRef: ReactInstance | null = null;

  handleButtonRef = (ref: any) => {
    this.buttonRef = ref;
  };

  handleClick = (event: MouseEvent) => {
    if (this.ripple.current && this.buttonRef) {
      this.ripple.current.startRipple(event, this.buttonRef);
    }
  };

  render() {
    const {
      children,
      icon,
      loading = false,
      type = 'default',
      size = 'md',
      ...props
    } = this.props;

    return (
      <StyledButton
        ref={this.handleButtonRef}
        onMouseUp={this.handleClick}
        hasIcon={Boolean(icon && !children)}
        loading={loading}
        size={size}
        type={type}
        {...props}
      >
        {!loading &&
          icon && (
            <Icon
              type={icon}
              size="20"
              mr={children ? 1 : 0}
              style={{ pointerEvents: 'none' }}
            />
          )}
        {children}
        <Ripple ref={this.ripple} />
      </StyledButton>
    );
  }
}

export default Button;
