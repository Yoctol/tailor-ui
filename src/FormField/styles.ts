import styled, { css } from 'styled-components';

export const Label = styled.label<{ required: boolean }>`
  display: block;
  margin-bottom: 2px;
  padding-left: 2px;
  color: ${(p) => p.theme.colors.gray700};
  font-size: ${(p) => p.theme.fontSizes.sm};
  letter-spacing: 0.2px;

  ${(p) =>
    p.required &&
    css`
      &::after {
        content: '*';
        margin-left: ${p.theme.space[1]};
      }
    `}
`;

export const ValidationMessage = styled.div`
  margin-top: 2px;
  padding-left: 2px;
  color: ${(p) => p.theme.colors.danger};
  font-size: ${(p) => p.theme.fontSizes.sm};
`;
