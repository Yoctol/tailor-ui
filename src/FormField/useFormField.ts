import { useContext, useEffect } from 'react';

import { tuplify } from '../utils';

import FormFieldContext from './FormFieldContext';

const useFormField = ({
  id,
  value,
  defaultValue,
}: {
  id?: string;
  value?: any;
  defaultValue?: any;
}) => {
  const { invalid, setValue, setLabelId, labelId } = useContext(
    FormFieldContext
  );

  useEffect(() => {
    if (id) {
      setLabelId(id);
    }
  }, [id, setLabelId]);

  useEffect(() => {
    const fieldValue = value || defaultValue;

    if (fieldValue) {
      setValue(fieldValue);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return tuplify(invalid, id || labelId, setValue);
};

export { useFormField };
