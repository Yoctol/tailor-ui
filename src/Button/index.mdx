---
name: Button
menu: General
route: /general/button
---

import { PropsTable, Playground } from 'docz';
import { MdCheck, MdDelete, MdClose, MdSearch } from 'react-icons/md';
import { Toggle } from 'react-powerplug';

import Box from '../Grid/Box';
import Icon from '../Icon';

import Button from './';

# Button

To trigger an operation.

## When To Use

A button means an operation (or a series of operations). Clicking a button will trigger corresponding business logic.

## Examples

```js
import { Button } from '@yoctol/ui';
```

### Basic

<Playground>
  <Button>Basic Button</Button>
  <Button text ml="2">Text Button</Button>
</Playground>

### Type

<Playground>
  <Button type="primary">Primary</Button>
  <Button type="secondary" ml="2">
    Secondary
  </Button>
  <Button type="info" ml="2">
    Info
  </Button>
  <Button type="success" ml="2">
    Success
  </Button>
  <Button type="warning" ml="2">
    Warning
  </Button>
  <Button type="danger" ml="2">
    Danger
  </Button>
</Playground>

### Rounded

<Playground>
  <Button type="primary" rounded>Primary</Button>
  <Button type="secondary" ml="2" rounded>
    Secondary
  </Button>
  <Button type="info" ml="2" rounded>
    Info
  </Button>
  <Button type="success" ml="2" rounded>
    Success
  </Button>
  <Button type="warning" ml="2" rounded>
    Warning
  </Button>
  <Button type="danger" ml="2" rounded>
    Danger
  </Button>
</Playground>

### Outlined

<Playground>
  <Button type="primary" outlined>Primary</Button>
  <Button type="secondary" ml="2" outlined>
    Secondary
  </Button>
  <Button type="info" ml="2" outlined>
    Info
  </Button>
  <Button type="success" ml="2" outlined>
    Success
  </Button>
  <Button type="warning" ml="2" outlined>
    Warning
  </Button>
  <Button type="danger" ml="2" outlined>
    Danger
  </Button>
</Playground>

### Rounded & Outlined

<Playground>
  <Button type="primary" rounded outlined>Primary</Button>
  <Button type="secondary" ml="2" rounded outlined>
    Secondary
  </Button>
  <Button type="info" ml="2" rounded outlined>
    Info
  </Button>
  <Button type="success" ml="2" rounded outlined>
    Success
  </Button>
  <Button type="warning" ml="2" rounded outlined>
    Warning
  </Button>
  <Button type="danger" ml="2" rounded outlined>
    Danger
  </Button>
</Playground>

### Icon

```js
import { MdCheck, MdDelete, MdClose, MdSearch } from 'react-icons/md';
```

<Playground>
  <Button type="primary" icon={MdCheck}>Confirm</Button>
  <Button type="danger" icon={MdDelete} ml="2">Delete</Button>
  <Button type="info" icon={MdSearch} ml="2">Search</Button>
  <br />
  <br />
  <Button type="primary" rounded icon={MdSearch} ml="2" />
  <Button type="danger" rounded icon={MdDelete} ml="2" />
  <Button type="info" rounded icon={MdSearch} ml="2" />
  <br />
  <br />
  <Button>
    <Icon type={MdCheck} fill="success" size="20" mr="2" />
    Success
  </Button>
  <Button ml="2">
    <Icon type={MdClose} fill="danger" size="20" mr="2" />
    Danger
  </Button>
</Playground>


### Disabled

<Playground>
  <Button type="primary" disabled>Primary</Button>
  <Button type="secondary" ml="2" disabled>
    Secondary
  </Button>
  <Button type="info" ml="2" disabled>
    Info
  </Button>
  <Button type="success" ml="2" disabled>
    Success
  </Button>
  <Button type="warning" ml="2" disabled>
    Warning
  </Button>
  <Button type="danger" ml="2" disabled>
    Danger
  </Button>
</Playground>

### Loading

<Playground>
  <Button type="primary" loading>Primary</Button>
  <Button type="secondary" ml="2" loading>
    Secondary
  </Button>
  <Button type="info" ml="2" loading>
    Info
  </Button>
  <Button type="success" ml="2" loading>
    Success
  </Button>
  <Button type="warning" ml="2" loading>
    Warning
  </Button>
  <Button type="danger" ml="2" loading>
    Danger
  </Button>
  <br />
  <br />
  <Toggle>
    {({ on, toggle }) => (
      <Button
        loading={on}
        onClick={() => {
          toggle()
          setTimeout(toggle, 3000)
        }}
      >
        Click me!
      </Button>
    )}
  </Toggle>
</Playground>

### Sizes

<Playground>
  <Button size="sm">
    Button sm
  </Button>

  <Button ml={2}>
    Button md
  </Button>

  <Button ml={2} size="lg">
    Button lg
  </Button>
</Playground>

### Block

<Playground>
  <Button block>
    Block Button
  </Button>
</Playground>

## API

<PropsTable of={Button} />
