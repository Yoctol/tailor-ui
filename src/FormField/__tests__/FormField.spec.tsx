import React, { useState } from 'react';
import userEvent from '@testing-library/user-event';
import { string } from 'yup';

import { render, screen } from 'test/test-utils';

import { Button } from '../../Button';
import { FormField } from '../FormField';
import { Input } from '../../Input';

describe('FormField', () => {
  it('should render correctly', () => {
    const { container } = render(
      <FormField label="Input">
        <Input id="input-id" placeholder="Placeholder" defaultValue="" />
      </FormField>
    );

    expect(screen.getByLabelText('Input')).toBeInTheDocument();
    expect(container.querySelector('#input-id')).toBeInTheDocument();
    expect(container.querySelector('label[for=input-id]')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render generated id correctly', () => {
    const { container } = render(
      <FormField label="Input">
        <Input placeholder="Placeholder" defaultValue="" />
      </FormField>
    );

    expect(screen.getByLabelText('Input')).toBeInTheDocument();
    expect(container.querySelector('#tailor_uid_1')).toBeInTheDocument();
    expect(
      container.querySelector('label[for=tailor_uid_1]')
    ).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('validationMessage', () => {
    it('should render FormField with error message correctly', async () => {
      const { container } = render(
        <FormField label="Input" validationMessage="Error Message">
          <Input placeholder="Placeholder" defaultValue="" />
        </FormField>
      );

      const message = screen.getByText('Error Message');

      expect(message).toBeVisible();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render FormField with error message correctly when change input value', async () => {
      const TextInput = () => {
        const [value, setValue] = useState('');

        return (
          <FormField
            label="Input"
            validationMessage={value === 'error' ? 'Error Message' : null}
          >
            <Input
              value={value}
              onChange={(e) => setValue(e.target.value)}
              placeholder="Placeholder"
            />
          </FormField>
        );
      };

      render(<TextInput />);
      const input = screen.getByPlaceholderText('Placeholder');

      expect(screen.queryByText('Error Message')).not.toBeInTheDocument();

      userEvent.type(input, 'error');

      const message = screen.getByText('Error Message');
      expect(message).toBeInTheDocument();
    });

    it('should render FormField with error message correctly when change validationMessage props', async () => {
      const TextInput = () => {
        const [hasError, setHasError] = useState(false);

        return (
          <>
            <Button onClick={() => setHasError(true)}>Show Error</Button>
            <FormField
              label="Input"
              validationMessage={hasError ? 'Error Message' : null}
            >
              <Input defaultValue="" />
            </FormField>
          </>
        );
      };

      render(<TextInput />);

      expect(screen.queryByText('Error Message')).not.toBeInTheDocument();

      userEvent.click(screen.getByText('Show Error'));

      const message = screen.getByText('Error Message');
      expect(message).toBeInTheDocument();
    });
  });

  describe('yup schema validator', () => {
    it('should render error message when input is invalid value', async () => {
      render(
        <FormField
          label="Input"
          validator={string().test(
            'is-error',
            'Error Message',
            (value) => value !== 'error'
          )}
        >
          <Input defaultValue="error" placeholder="Placeholder" />
        </FormField>
      );

      const message = screen.getByText('Error Message');
      expect(message).toBeInTheDocument();
    });

    it('should not render error message when input is valid value', () => {
      render(
        <FormField
          label="Input"
          validator={string().test(
            'is-error',
            'Error Message',
            (value) => value !== 'error'
          )}
        >
          <Input defaultValue="valid input" placeholder="Placeholder" />
        </FormField>
      );

      expect(screen.queryByText('Error Message')).not.toBeInTheDocument();
    });

    it('should render FormField with error message correctly when change input value', async () => {
      render(
        <FormField
          label="Input"
          validator={string().test(
            'is-error',
            'Error Message',
            (value) => value !== 'error'
          )}
        >
          <Input defaultValue="valid input" placeholder="Placeholder" />
        </FormField>
      );

      const input = screen.getByPlaceholderText('Placeholder');

      expect(screen.queryByText('Error Message')).not.toBeInTheDocument();
      userEvent.type(input, '{selectall}{backspace}error');

      const message = screen.getByText('Error Message');
      expect(message).toBeInTheDocument();
    });
  });

  describe('function validator', () => {
    it('should render error message when input is invalid value', async () => {
      render(
        <FormField
          label="Input"
          validator={(value) => (value === 'error' ? 'Error Message' : null)}
        >
          <Input defaultValue="error" placeholder="Placeholder" />
        </FormField>
      );

      const message = screen.getByText('Error Message');
      expect(message).toBeInTheDocument();
    });

    it('should not render error message when input is valid value', () => {
      render(
        <FormField
          label="Input"
          validator={(value) => (value === 'error' ? 'Error Message' : null)}
        >
          <Input defaultValue="valid input" placeholder="Placeholder" />
        </FormField>
      );

      expect(screen.queryByText('Error Message')).not.toBeInTheDocument();
    });

    it('should render FormField with error message correctly when change input value', async () => {
      render(
        <FormField
          label="Input"
          validator={(value) => (value === 'error' ? 'Error Message' : null)}
        >
          <Input defaultValue="valid input" placeholder="Placeholder" />
        </FormField>
      );

      const input = screen.getByPlaceholderText('Placeholder');

      expect(screen.queryByText('Error Message')).not.toBeInTheDocument();
      userEvent.type(input, '{selectall}{backspace}error');

      const message = screen.getByText('Error Message');
      expect(message).toBeInTheDocument();
    });
  });

  describe('object validator', () => {
    it('should render error message when input is invalid value', async () => {
      render(
        <FormField
          label="Input"
          validator={{
            rule: (value) => value === 'error1',
            message: 'error message',
          }}
        >
          <Input defaultValue="error1" placeholder="Placeholder" />
        </FormField>
      );

      const message = screen.getByText('error message');
      expect(message).toBeInTheDocument();
    });

    it('should render error message when input is invalid value', async () => {
      render(
        <FormField
          label="Input"
          validator={[
            {
              rule: (value) => value === 'error1',
              message: 'error message',
            },
            {
              rule: (value) => value === 'error2',
              message: 'another error message',
            },
          ]}
        >
          <Input defaultValue="error1" placeholder="Placeholder" />
        </FormField>
      );

      const message = screen.getByText('error message');
      expect(message).toBeInTheDocument();
    });

    it('should render error message when input is another invalid value', async () => {
      render(
        <FormField
          label="Input"
          validator={[
            {
              rule: (value) => value === 'error1',
              message: 'error message',
            },
            {
              rule: (value) => value === 'error2',
              message: 'another error message',
            },
          ]}
        >
          <Input defaultValue="error2" placeholder="Placeholder" />
        </FormField>
      );

      const message = screen.getByText('another error message');
      expect(message).toBeInTheDocument();
    });

    it('should not render error message when input is valid', () => {
      render(
        <FormField
          label="Input"
          validator={[
            {
              rule: (value) => value === 'error1',
              message: 'error message',
            },
            {
              rule: (value) => value === 'error2',
              message: 'another error message',
            },
          ]}
        >
          <Input defaultValue="valid input" placeholder="Placeholder" />
        </FormField>
      );

      expect(screen.queryByText('error message')).not.toBeInTheDocument();
      expect(
        screen.queryByText('another error message')
      ).not.toBeInTheDocument();
    });

    it('should render FormField with error message correctly when change input value', async () => {
      render(
        <FormField
          label="Input"
          validator={[
            {
              rule: (value) => value === 'error1',
              message: 'error message',
            },
            {
              rule: (value) => value === 'error2',
              message: 'another error message',
            },
          ]}
        >
          <Input defaultValue="valid input" placeholder="Placeholder" />
        </FormField>
      );

      expect(screen.queryByText('Error Message')).not.toBeInTheDocument();

      const input = screen.getByPlaceholderText('Placeholder');

      userEvent.type(input, '{selectall}{backspace}error1');
      const message = screen.getByText('error message');
      expect(message).toBeInTheDocument();

      userEvent.type(input, '{selectall}{backspace}error2');
      const message2 = screen.getByText('another error message');
      expect(message2).toBeInTheDocument();
    });
  });
});
