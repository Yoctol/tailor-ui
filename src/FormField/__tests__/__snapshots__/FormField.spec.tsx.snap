// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField should render correctly 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 16px;
}

.c1 {
  display: block;
  margin-bottom: 2px;
  padding-left: 2px;
  color: #283040;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
}

.c2 {
  display: block;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #97a4ba;
  outline: none;
  background-color: #ffffff;
  color: #283040;
  line-height: 1.5;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  height: 2.25rem;
  padding: 4px 0.75rem;
  font-size: 1rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c2:hover {
  border-color: #328ef5;
}

.c2:focus {
  border-color: #328ef5;
  box-shadow: inset 0 0 0 2px #eef1f8;
}

.c2:disabled,
.c2 [disabled] {
  border-color: #97a4ba;
  background-color: #c4cfdd;
  color: #67758d;
  cursor: not-allowed;
}

.c2::-webkit-input-placeholder {
  color: #97a4ba;
}

.c2::-moz-placeholder {
  color: #97a4ba;
}

.c2:-ms-input-placeholder {
  color: #97a4ba;
}

.c2::placeholder {
  color: #97a4ba;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="input-id"
  >
    Input
  </label>
  <input
    class="c2"
    id="input-id"
    placeholder="Placeholder"
    value=""
  />
</div>
`;

exports[`FormField should render generated id correctly 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 16px;
}

.c1 {
  display: block;
  margin-bottom: 2px;
  padding-left: 2px;
  color: #283040;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
}

.c2 {
  display: block;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #97a4ba;
  outline: none;
  background-color: #ffffff;
  color: #283040;
  line-height: 1.5;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  height: 2.25rem;
  padding: 4px 0.75rem;
  font-size: 1rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c2:hover {
  border-color: #328ef5;
}

.c2:focus {
  border-color: #328ef5;
  box-shadow: inset 0 0 0 2px #eef1f8;
}

.c2:disabled,
.c2 [disabled] {
  border-color: #97a4ba;
  background-color: #c4cfdd;
  color: #67758d;
  cursor: not-allowed;
}

.c2::-webkit-input-placeholder {
  color: #97a4ba;
}

.c2::-moz-placeholder {
  color: #97a4ba;
}

.c2:-ms-input-placeholder {
  color: #97a4ba;
}

.c2::placeholder {
  color: #97a4ba;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="tailor_uid_1"
  >
    Input
  </label>
  <input
    class="c2"
    id="tailor_uid_1"
    placeholder="Placeholder"
    value=""
  />
</div>
`;

exports[`FormField validationMessage should render FormField with error message correctly 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 16px;
}

.c1 {
  display: block;
  margin-bottom: 2px;
  padding-left: 2px;
  color: #283040;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
}

.c3 {
  margin-top: 2px;
  padding-left: 2px;
  color: #ff5757;
  font-size: 0.875rem;
}

.c2 {
  display: block;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #97a4ba;
  outline: none;
  background-color: #ffffff;
  color: #283040;
  line-height: 1.5;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border-color: #ff5757 !important;
  box-shadow: none !important;
  height: 2.25rem;
  padding: 4px 0.75rem;
  font-size: 1rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c2:hover {
  border-color: #328ef5;
}

.c2:focus {
  border-color: #328ef5;
  box-shadow: inset 0 0 0 2px #eef1f8;
}

.c2:disabled,
.c2 [disabled] {
  border-color: #97a4ba;
  background-color: #c4cfdd;
  color: #67758d;
  cursor: not-allowed;
}

.c2::-webkit-input-placeholder {
  color: #97a4ba;
}

.c2::-moz-placeholder {
  color: #97a4ba;
}

.c2:-ms-input-placeholder {
  color: #97a4ba;
}

.c2::placeholder {
  color: #97a4ba;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="tailor_uid_1"
  >
    Input
  </label>
  <input
    class="c2"
    id="tailor_uid_1"
    placeholder="Placeholder"
    value=""
  />
  <div
    style="height: 22px; opacity: 1;"
  >
    <div
      class="c3"
    >
      Error Message
    </div>
  </div>
</div>
`;
